clover.rb

```diff
@@ -342,19 +348,41 @@ def assets
     password_confirm_label "Password Confirmation"
     before_create_account do
       Validation.validate_cloudflare_turnstile(param("cf-turnstile-response"))
-      account[:id] = Account.generate_uuid
-      account[:name] = param("name")
-      Validation.validate_account_name(account[:name])
+      scope.before_rodauth_create_account(account, param("name"))
     end
     after_create_account do
-      account = Account[account_id]
-      account.create_project_with_default_policy("Default")
-      ProjectInvitation.where(email: account.email).each do |inv|
-        account.associate_with_project(inv.project)
-        if (managed_policy = Authorization::ManagedPolicy.from_name(inv.policy))
-          managed_policy.apply(inv.project, [account], append: true)
+      scope.after_rodauth_create_account(account_id)
+    end
+
+    # :nocov:
+    if Config.omniauth_github_id
+      require "omniauth-github"
+      omniauth_provider :github, Config.omniauth_github_id, Config.omniauth_github_secret
+    end
+    if Config.omniauth_google_id
+      require "omniauth-google-oauth2"
+      omniauth_provider :google_oauth2, Config.omniauth_google_id, Config.omniauth_google_secret, name: :google
+    end
+    # :nocov:
+
+    before_omniauth_create_account do
+      scope.before_rodauth_create_account(account, omniauth_name)
+    end
+
+    after_omniauth_create_account do
+      scope.after_rodauth_create_account(account_id)
+    end
+
+    before_omniauth_callback_route do
+      account = Account[account_from_omniauth&.dig(:id)]
+      if authenticated?
+        if (account && account.id != scope.current_account.id) || !omniauth_email.casecmp?(scope.current_account.email)
+          flash["error"] = "Your account's email address is different from the email address associated with the #{omniauth_provider.capitalize} account."
```

Comment:
If users are manually linking accounts and we are not linking them automatically, does it matter if the email address for the account is different?

Reply:
The problem is that `rodauth-omniauth` initially attempts to find the account using the provider and UID, then tries using email. If it can't find an account, it simply creates a new one with the new email and logs into it. As a result, you might unintentionally switch to a different account.

https://github.com/janko/rodauth-omniauth/blob/3e8b627b8b3d21cc382f4a9d421fceef483b5d72/lib/rodauth/features/omniauth.rb#L98

I can override some methods, and for authenticated users, we can use the current user's account ID instead of creating a new one. However, to be on the safe side, I decided not to allow different emails. For example, I shouldn't connect my personal Gmail account to my Ubicloud account with my business email.


I've simplified this check a bit. Since `account_from_omniauth` finds the account by email, there's no need to separately check the email. If the account isn't found, it means the emails are different.

```
# before
if (account && account.id != scope.current_account.id) || !omniauth_email.casecmp?(scope.current_account.email)

# after
unless account && account.id == scope.current_account.id
```

I also added `omniauth_create_account? { !authenticated? }`. So it doesn't allow to create a new account accidentally, for `authenticated?` users.